pkg/kademlia/testrouting/testrouting.go

```diff
@@ -0,0 +1,299 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package testrouting
+
+import (
+	"sort"
+	"sync"
+	"time"
+
+	"storj.io/storj/pkg/dht"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/storage"
+)
+
+type nodeData struct {
+	node      *pb.Node
+	ordering  int64
+	timestamp time.Time
+	fails     int
+	inCache   bool
+}
+
+// Table is a routing table that tries to be as correct as possible at
+// the expense of performance.
+type Table struct {
+	self            storj.NodeID
+	bucketSize      int
+	cacheSize       int
+	allowedFailures int
+
+	mu      sync.Mutex
+	counter int64
+	nodes   map[storj.NodeID]*nodeData
+}
+
+// New creates a new Table. self is the owning node's node id, bucketSize is
+// the kademlia k value, cacheSize is the size of each bucket's replacement
+// cache, and allowedFailures is the number of failures on a given node before
+// the node is removed from the table.
+func New(self storj.NodeID, bucketSize, cacheSize, allowedFailures int) *Table {
+	return &Table{
+		self:            self,
+		bucketSize:      bucketSize,
+		cacheSize:       cacheSize,
+		allowedFailures: allowedFailures,
+		nodes:           map[storj.NodeID]*nodeData{},
+	}
+}
+
+// make sure the Table implements the right interface
+var _ dht.RoutingTable = (*Table)(nil)
+
+// K returns the Table's routing depth, or Kademlia k value
+func (t *Table) K() int { return t.bucketSize }
+
+// CacheSize returns the size of
+func (t *Table) CacheSize() int { return t.cacheSize }
+
+// ConnectionSuccess should be called whenever a node is successfully connected
+// to. It will add or update the node's entry in the routing table.
+func (t *Table) ConnectionSuccess(node *pb.Node) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// don't add ourselves
+	if node.Id == t.self {
+		return nil
+	}
+
+	// if the node is already here, update it
+	if cell, exists := t.nodes[node.Id]; exists {
+		cell.node = node
+		cell.timestamp = time.Now()
+		cell.fails = 0
+		// skip placement order and cache status
+		return nil
+	}
+
+	// add unconditionally (it might be going into a replacement cache)
+	t.nodes[node.Id] = &nodeData{
+		node:      node,
+		ordering:  t.counter,
+		timestamp: time.Now(),
```

Comment:
Thinking: I wonder whether using logical time here might make more sense?

Reply:
that's what the ordering field is